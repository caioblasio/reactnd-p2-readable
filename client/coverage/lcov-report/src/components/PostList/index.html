<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/components/PostList</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> src/components/PostList
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">7.69% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>5/65</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/33</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/27</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">8.93% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>5/56</span>
      </div>
    </div>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="Headline.js"><a href="Headline.js.html">Headline.js</a></td>
	<td data-value="9.09" class="pic low"><div class="chart"><div class="cover-fill" style="width: 9%;"></div><div class="cover-empty" style="width:91%;"></div></div></td>
	<td data-value="9.09" class="pct low">9.09%</td>
	<td data-value="11" class="abs low">1/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="12.5" class="pct low">12.5%</td>
	<td data-value="8" class="abs low">1/8</td>
	</tr>

<tr>
	<td class="file low" data-value="Post.js"><a href="Post.js.html">Post.js</a></td>
	<td data-value="6.67" class="pic low"><div class="chart"><div class="cover-fill" style="width: 6%;"></div><div class="cover-empty" style="width:94%;"></div></div></td>
	<td data-value="6.67" class="pct low">6.67%</td>
	<td data-value="15" class="abs low">1/15</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="5" class="abs low">0/5</td>
	<td data-value="7.14" class="pct low">7.14%</td>
	<td data-value="14" class="abs low">1/14</td>
	</tr>

<tr>
	<td class="file low" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="7.69" class="pic low"><div class="chart"><div class="cover-fill" style="width: 7%;"></div><div class="cover-empty" style="width:93%;"></div></div></td>
	<td data-value="7.69" class="pct low">7.69%</td>
	<td data-value="39" class="abs low">3/39</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="21" class="abs low">0/21</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="16" class="abs low">0/16</td>
	<td data-value="8.82" class="pct low">8.82%</td>
	<td data-value="34" class="abs low">3/34</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Nov 22 2018 23:35:45 GMT-0200 (-02)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
