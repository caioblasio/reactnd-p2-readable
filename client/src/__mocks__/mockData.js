const testPosts = {

  "8xf0y6ziyjabvozdd253nd": {
    id: '8xf0y6ziyjabvozdd253nd',
    timestamp: 1467166872634,
    title: 'Udacity is the best place to learn React',
    body: 'Everyone says so after all.',
    author: 'thingtwo',
    category: 'categoryone',
    voteScore: 6,
    deleted: false,
    commentCount: 2
  },
  "6ni6ok3ym7mf1p33lnez": {
    id: '6ni6ok3ym7mf1p33lnez',
    timestamp: 1468479767190,
    title: 'Learn Redux in 10 minutes!',
    body: 'Just kidding. It takes more than 10 minutes to learn technology.',
    author: 'thingone',
    category: 'categorytwo',
    voteScore: -5,
    deleted: false,
    commentCount: 0
  }
}

const testCategories = [
  {
    name: 'Category One',
    path: 'categoryone'
  },
  {
    name: 'Category Two',
    path: 'categorytwo'
  }
]

const testComments = [
  {
    id: '894tuq4ut84ut8v4t8wun89g',
    parentId: "8xf0y6ziyjabvozdd253nd",
    timestamp: 1468166872634,
    body: 'Hi there! I am a COMMENT.',
    author: 'thingtwo',
    voteScore: 6,
    deleted: false,
    parentDeleted: false
  },
  {
    id: '8tu4bsun805n8un48ve89',
    parentId: "8xf0y6ziyjabvozdd253nd",
    timestamp: 1469479767190,
    body: 'Comments. Are. Cool.',
    author: 'thingone',
    voteScore: -5,
    deleted: false,
    parentDeleted: false
  }
  
]

/* JSON headers for API requests */
const jsonHeaders = new Headers({
	Accept: 'application/json',
	'Content-Type': 'application/json'
});

export { testPosts, testCategories, testComments, jsonHeaders };