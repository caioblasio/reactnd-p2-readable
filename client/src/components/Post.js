import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { getFormatedDate } from '../utils/date';
import { withStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import CardActionArea from '@material-ui/core/CardActionArea';
import IconButton from '@material-ui/core/IconButton';
import Icon from '@material-ui/core/Icon';
import Typography from '@material-ui/core/Typography';
import ThumbUp from '@material-ui/icons/ThumbUp';
import ThumbDown from '@material-ui/icons/ThumbDown';
import Comment from '@material-ui/icons/Comment';
import Divider from '@material-ui/core/Divider';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import MoreVertIcon from '@material-ui/icons/MoreVert';

import { addVoteToPost } from '../actions/posts';

import VoteControl from './VoteControl';

const styles = theme => ({
  card: {
    width: '100%',
    maxWidth: 380,
    margin: theme.spacing.unit * 2,
    display: 'flex',
    flexDirection: 'column',
  },
  link: {
    ...theme.link
  },
  actions: {
    display: 'flex',
    alignItems: 'center',
  },
  actionIcon: {
    display: 'flex',
    alignItems: 'center',
  },
  footer: {
    marginTop: 'auto',
  }
});



class Post extends Component {

  state = {
    anchorEl: null,
  };

  handleMenuClick = event => {
    this.setState({ anchorEl: event.currentTarget });
  };

  handleMenuClose = () => {
    this.setState({ anchorEl: null });
  };

  render() {
    const { post, classes } = this.props
    const { author, category, body, commentCount, id, timestamp, title, voteScore } = post
    const { anchorEl } = this.state;
    const open = Boolean(anchorEl);

    return (
      <Card className={classes.card}>
        <Menu
          id="long-menu"
          anchorEl={anchorEl}
          open={open}
          onClose={this.handleMenuClose}
        >
          <MenuItem onClick={this.handleMenuClose}>
            Edit
          </MenuItem>
          <MenuItem onClick={this.handleMenuClose}>
            Remove
          </MenuItem>
        </Menu>
        <Link to={`/${category}/${id}`} className={classes.link}>
          <CardHeader
            action={
              <IconButton
                aria-label="More"
                aria-owns={open ? 'long-menu' : null}
                aria-haspopup="true"
                onClick={this.handleMenuClick}
              >
              <MoreVertIcon />
            </IconButton>
            }
            title={title}
            subheader={`by ${author} ${getFormatedDate(timestamp)}`}
          />
          <CardContent>
            <Typography component="p">
              {`${body.substring(0,160)}...`}
            </Typography>
          </CardContent>
        </Link>
        <div className={classes.footer}>
          <Divider />
          <CardActions className={classes.actions} disableActionSpacing>
            <div className={classes.actionIcon} style={{flexGrow: 1}}>
              <VoteControl
                type="post"
                voteScore={voteScore}
                id={id}
              />
            </div>
            <div className={classes.actionIcon} >
              <IconButton
                aria-label="Comments"
              >
                <Comment />
              </IconButton>
              <Typography component="p">
                {commentCount}
              </Typography>
            </div>
          </CardActions>
        </div>
      </Card>
    )
  }
}

function mapStateToProps({ posts }, { id }) {
  const post = posts[id]
  return {
    post: post
  }
}

export default withStyles(styles)(connect(mapStateToProps)(Post))