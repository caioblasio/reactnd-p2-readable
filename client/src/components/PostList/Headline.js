import React, { Component, Fragment } from 'react'
import { withStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import Divider from '@material-ui/core/Divider';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import SortIcon from '@material-ui/icons/Sort';

const styles = theme => ({
  root: {
    display: 'flex',
    paddingBottom: theme.spacing.unit,
  },
  title: {
    flexGrow: 1,
  },
  header: {
    color: theme.palette.text.secondary,
  },
  leftIcon: {
    marginRight: theme.spacing.unit,
  },
  center: {
    textAlign: 'center',
  }
});

/**
 * @description Headline component for a post list
 * @param {string} title
 * @param {bool} side
 * @param {function()} onSort
 * @param {string} sortOption
 * @param {object} classes
*/
class Headline extends Component {

  static propType = {
    title: PropTypes.string,
    side: PropTypes.bool,
    onSort: PropTypes.func.isRequired,
    sortOption: PropTypes.string.isRequired,
    classes: PropTypes.object.isRequired
  };

  state = {
    anchorEl: null,
  };

  /**
   * @description Handles click on sort select
   * @param {object} 
  */
  handleClick = event => {
    this.setState({ anchorEl: event.currentTarget });
  };

  /**
   * @description Handles close on sort select
  */
  handleClose = () => {
    this.setState({ anchorEl: null });
  };


  render() {
    const { anchorEl } = this.state,
     { title, onSort, sortOption, side, classes } = this.props;

    return (
        <Fragment>
          <div className={classes.root}>
            <div className={`${classes.title} ${side ? classes.center : ''}`}>
              <Typography variant="h5" className={classes.header}>
                {title}
              </Typography>
            </div>
            {!side &&
              <div>
                <Button 
                  variant="contained" 
                  size="small" 
                  className={classes.button}
                  aria-owns={anchorEl ? 'simple-menu' : undefined}
                  aria-haspopup="true"
                  onClick={this.handleClick}
                >
                  <SortIcon className={classes.leftIcon} />
                  {`Sorted by ${sortOption}`}
                </Button>
                <Menu
                  id="sort-options"
                  anchorEl={anchorEl}
                  open={Boolean(anchorEl)}
                  onClose={this.handleClose}
                >
                  <MenuItem
                    selected={sortOption === 'score'}
                    onClick={() => {onSort('score'); this.handleClose()}}
                  >
                    Vote Score
                  </MenuItem>
                  <MenuItem 
                    selected={sortOption === 'date'}
                    onClick={() => {onSort('date'); this.handleClose();}}
                  >
                    Date
                  </MenuItem>
                </Menu>
              </div>
            }
          </div>
          {!side && <Divider/>}
        </Fragment>
    )
  }
}

export default withStyles(styles)(Headline)