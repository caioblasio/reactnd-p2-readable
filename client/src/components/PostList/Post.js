import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { getFormatedDate } from '../../utils/date';
import { withStyles } from '@material-ui/core/styles';
import VoteControl from '../VoteControl';
import FavoriteControl from '../FavoriteControl';
import PropTypes from 'prop-types';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import Comment from '@material-ui/icons/Comment';
import Divider from '@material-ui/core/Divider';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import MoreVertIcon from '@material-ui/icons/MoreVert';

const styles = theme => ({
  card: {
    width: '100%',
    maxWidth: 380,
    margin: theme.spacing.unit * 2,
    display: 'flex',
    flexDirection: 'column',
  },
  link: {
    ...theme.link,
    outline: 0,
  },
  actions: {
    display: 'flex',
    alignItems: 'center',
  },
  actionIcon: {
    display: 'flex',
    alignItems: 'center',
  },
  footer: {
    marginTop: 'auto',
    paddingRight: theme.spacing.unit,
  }
});

/**
 * @description Single post card component
 * @param {object} post
 * @param {function()} onDelete
 * @param {object} classes
*/
class Post extends Component {

  static propType = {
    post: PropTypes.object.isRequired,
    onDelete: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired
  };

  state = {
    anchorEl: null,
  };

  /**
   * @description Handles click on post menu
   * @param {object} event
  */
  handleMenuClick = event => {
    event.preventDefault();
    this.setState({ anchorEl: event.currentTarget });
  };

  /**
   * @description Handles close on post menu
  */
  handleMenuClose = () => {
    this.setState({ anchorEl: null });
  };

  /**
   * @description Handles deletion of a post
  */
  handleRemovePost = () =>{
     const { onDelete, post } = this.props;
     const alertConfirmation = window.confirm(
        `Are you sure you want to delete the post '${post.title}'?`
      )
    if (alertConfirmation) {
      onDelete(post.id)
    }

    this.handleMenuClose();
  }

  render() {
    const { post, classes } = this.props
    const { author, category, body, commentCount, id, timestamp, title, voteScore } = post
    const { anchorEl } = this.state;
    const open = Boolean(anchorEl);

    return (
      <Card className={`${classes.card} post`}>
        <Menu
          id="long-menu"
          anchorEl={anchorEl}
          open={open}
          onClose={this.handleMenuClose}
        >
        <Link
          className={classes.link}
          to={`/edit/${id}`}
          >
            <MenuItem onClick={this.handleMenuClose}>Edit</MenuItem>
          </Link>
          <MenuItem onClick={this.handleRemovePost}>
            Remove
          </MenuItem>
        </Menu>
        <Link to={`/${category}/${id}`} className={classes.link}>
          <CardHeader
            action={
              <IconButton
                aria-label="More"
                aria-owns={open ? 'long-menu' : null}
                aria-haspopup="true"
                onClick={this.handleMenuClick}
              >
              <MoreVertIcon />
            </IconButton>
            }
            title={title}
            subheader={`by ${author} ${getFormatedDate(timestamp)} ${category}`}
          />
          <CardContent>
            <Typography component="p">
              {`${body.substring(0,160)}...`}
            </Typography>
          </CardContent>
        </Link>
        <div className={classes.footer}>
          <Divider />
          <CardActions className={classes.actions} disableActionSpacing>
            <div className={classes.actionIcon} style={{flexGrow: 1}}>
              <VoteControl
                type="post"
                voteScore={voteScore}
                id={id}
              />
            </div>
            <div className={classes.actionIcon} >
              <FavoriteControl
                id={id}
              />
              <IconButton
                aria-label="Comments"
              >
                <Comment />
              </IconButton>
              <Typography component="p">
                {commentCount}
              </Typography>
            </div>
          </CardActions>
        </div>
      </Card>
    )
  }
}

export default withStyles(styles)(Post)