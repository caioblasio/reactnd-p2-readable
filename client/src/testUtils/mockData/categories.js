export const testCategories = [
  {
    name: 'Category One',
    path: 'categoryone'
  },
  {
    name: 'Category Two',
    path: 'categorytwo'
  }
]