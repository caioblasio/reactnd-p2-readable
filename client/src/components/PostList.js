import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles';
import Post from './Post';

import sortList from '../utils/sort';
import { fetchPosts, fetchPostsByCategory } from '../actions/posts';
import CircularProgress from '@material-ui/core/CircularProgress';

const styles = theme => ({
  list: {
    display: 'flex',
    flexWrap: 'wrap',

  },
  loading: {
    marginTop: theme.spacing.unit * 4,
    textAlign: 'center',
  }
});

class PostList extends Component {

  componentDidMount(){
    const { category, dispatch } = this.props;
    if(category){
      dispatch(fetchPostsByCategory(category));
    } else {
      dispatch(fetchPosts());
    }
  }

  componentDidUpdate(prevProps){
    const { category } = this.props
    const prevCategory = prevProps.category

    if (prevCategory !== category) {
      this.props.dispatch(fetchPostsByCategory(category));
    }
  }
  
  render() {
    const { postsIds, isLoading, classes } = this.props
    console.log('sorted', postsIds)
    return (
      <Fragment>
      {!isLoading && 
      <div className={classes.list}>
        {postsIds.map(postId => (
          <Post key={postId} id={postId} />
        ))}
      </div>}
      {isLoading && 
        <div className={classes.loading}>
          <CircularProgress />
        </div>
      }
      </Fragment>
    )
  }
}

function mapStateToProps({ posts, loading, sort }, { category, excludeId }) {
  
  const sortedPosts = sortList(Object.keys(posts).map(key => posts[key]), sort)
    .reduce((acc, post) => ({ ...acc, [post.id]: post}), {});

  const postIdsArray = !excludeId
    ? Object.keys(sortedPosts) 
    : Object.keys(sortedPosts).filter(postId => postId !== excludeId)

  return {
    postsIds: category
      ? postIdsArray.filter(id => (
          posts[id].category === category
        ))
      : postIdsArray,
      isLoading: loading,
  }
}

export default withStyles(styles)(connect(mapStateToProps)(PostList))