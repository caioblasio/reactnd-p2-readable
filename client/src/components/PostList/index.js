import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles';
import Post from './Post';
import Headline from './Headline';
import sortList from '../../utils/sort';
import { changeSort } from '../../actions/sort';
import { fetchPosts, fetchPostsByCategory, removePost } from '../../actions/posts';

const styles = theme => ({
  list: {
    display: 'flex',
    flexWrap: 'wrap',

  },
  loading: {
    marginTop: theme.spacing.unit * 4,
    textAlign: 'center',
  }
});

class PostList extends Component {

  componentDidMount(){
    const { category, fetchPosts, fetchPostsByCategory } = this.props;

    if(category){
      fetchPostsByCategory(category);
    } else {
      fetchPosts();
    }
  }

  componentDidUpdate(prevProps){
    const { category, fetchPostsByCategory } = this.props
    const prevCategory = prevProps.category

    if (prevCategory !== category) {
      fetchPostsByCategory(category);
    }
  }
  
  render() {
    const { posts, deletePost, handleSort, sort, headline, classes } = this.props

    return (
      <Fragment>
        <Headline title={headline} onSort={handleSort} sortOption={sort}/>
        <div className={classes.list}>
          {posts.map(post => (
            <Post key={post.id} post={post} onDelete={deletePost}/>
          ))}
        </div>
      </Fragment>
    )
  }
}

const mapStateToProps = ({ posts, sort }, { category, excludeId, includeOnlyIds }) => {
  
  let sortedPosts = sortList(Object.keys(posts).map(key => posts[key]), sort);

  if(category){
    sortedPosts = sortedPosts.filter(post => post.category === category)
  }

  if(excludeId){
    sortedPosts = sortedPosts.filter(post => post.id !== excludeId)
  }

  if(includeOnlyIds){
    sortedPosts = sortedPosts.filter(post => includeOnlyIds.includes(post.id))
  }

  return {
    posts: sortedPosts,
    sort
  }
}


const mapDispatchToProps = dispatch => {
  return {
    fetchPosts: () => {
      dispatch(fetchPosts())
    },
    fetchPostsByCategory: (category) => {
      dispatch(fetchPostsByCategory(category))
    },
    deletePost: (id) => {
      dispatch(removePost(id))
    },
    handleSort: (value) => {
      dispatch(changeSort(value))
    }
  }
}

export default withStyles(styles)(connect(mapStateToProps, mapDispatchToProps)(PostList))