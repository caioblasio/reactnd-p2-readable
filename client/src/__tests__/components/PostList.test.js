import React from 'react';
import PostList from '../../components/PostList';
import Post from '../../components/PostList/Post';
import Headline from '../../components/PostList/Headline';
import configureStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';
import { testPosts, testComments, jsonHeaders } from '../../__mocks__/mockData';
import { MemoryRouter, Router } from 'react-router-dom';
import thunk from 'redux-thunk'
import IconButton from '@material-ui/core/IconButton';

import { createShallow } from '@material-ui/core/test-utils';
import { MenuItem } from '@material-ui/core';

const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

describe('<PostList />', () => {

  let wrapper, store;

  beforeAll(() => {
    const initialState = {
      posts: testPosts,
      comments: testComments,
      loading: false
    }
    store = mockStore(initialState);
  });

  beforeEach(() => {
    fetch.resetMocks();
  });

  it('renders all posts', () => {

    fetch.mockResponse(JSON.stringify(testPosts), { jsonHeaders })

    wrapper = shallow(
      <PostList
        store={store}
      />
    );
    const component = wrapper.dive();
    expect(component.dive().find(Post)).toHaveLength(2);
  });

  it('renders only posts by category posts', () => {

    fetch.mockResponse(JSON.stringify(testPosts), { jsonHeaders })

    wrapper = shallow(
      <PostList
        store={store}
        category="categoryone"
      />
    );
    const component = wrapper.dive();
    expect(component.dive().find(Post)).toHaveLength(1);
  });

  it('renders only specified number of posts', () => {

    fetch
      .mockResponse(JSON.stringify(testPosts), { jsonHeaders })

    wrapper = shallow(
      <PostList
        store={store}
        category="categoryone"
        maxQty={1}
      />
    );
    const component = wrapper.dive();
    expect(component.dive().find(Post)).toHaveLength(1);
  });

  it('does not render excluded post', () => {

    const excludeId = testPosts[Object.keys(testPosts)[0]].id

    fetch
      .mockResponse(JSON.stringify(testPosts), { jsonHeaders })

    wrapper = shallow(
      <PostList
        store={store}
        excludeId={excludeId}
      />
    );
    const component = wrapper.dive();
    expect(component.dive().find(Post).props().post.id).not.toEqual(excludeId);
    expect(component.dive().find(Post)).toHaveLength(1);

  });

  it('renders only specified posts', () => {

    const includeOnlyId = testPosts[Object.keys(testPosts)[0]].id

    fetch
      .mockResponse(JSON.stringify(testPosts), { jsonHeaders })

    wrapper = shallow(
      <PostList
        store={store}
        includeOnlyIds={includeOnlyId}
      />
    );
    const component = wrapper.dive();
    expect(component.dive().find(Post).props().post.id).toEqual(includeOnlyId);
    expect(component.dive().find(Post)).toHaveLength(1);

  });

  it('renders only searched posts', () => {

    const initialState = {
      posts: testPosts,
      comments: testComments,
      search: 'Udacity is the best place',
      loading: false
    }

    let store = mockStore(initialState);
    fetch.mockResponse(JSON.stringify(testPosts), { jsonHeaders })
    wrapper = shallow(
      <PostList
        store={store}
      />
    );
    const component = wrapper.dive();
    expect(component.dive().find(Post).props().post.title).toEqual('Udacity is the best place to learn React');
    expect(component.dive().find(Post)).toHaveLength(1);

  });


  it('renders posts from a category after all posts are rendered', () => {

    fetch.mockResponse(JSON.stringify(testPosts), { jsonHeaders })
    wrapper = shallow(
      <PostList
        store={store}
      />
    );
    const component = wrapper.dive();
    expect(component.dive().find(Post)).toHaveLength(2);
    component.setProps({ category: "categoryone" })
    expect(component.dive().find(Post)).toHaveLength(1);

  });
});

describe('<Post />', () => {

  let wrapper;

  it('renders a posts', () => {

    let onDelete = jest.fn();
    const post = testPosts[Object.keys(testPosts)[0]]

    wrapper = shallow(
      <Post
        post={post}
        onDelete={onDelete}
      />
    );
    const component = wrapper.dive();
  
    expect(toJson(component.dive())).toMatchSnapshot();
  });

  it('handle delete post', () => {

    let onDelete = jest.fn();
    const post = testPosts[Object.keys(testPosts)[0]]
    window.confirm = jest.fn(() => true);

    wrapper = shallow(
      <Post
        post={post}
        onDelete={onDelete}
      />
    );
    const component = wrapper.dive();
  
    component.dive().find(MenuItem).last().simulate('click');
    expect(onDelete.mock.calls).toHaveLength(1);
  });

});

describe('<Headline />', () => {

  let wrapper;

  it('renders a Headline', () => {

    let onSort = jest.fn();

    wrapper = shallow(
      <Headline
        title="All Posts"
        onSort={onSort}
        sortOption="score"
      />
    );
    const component = wrapper.dive();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('calls sort method with date option', () => {

    let onSort = jest.fn();

    wrapper = shallow(
      <Headline
        title="All Posts"
        onSort={onSort}
        sortOption="score"
      />
    );
    const component = wrapper.dive();
    component.find(MenuItem).last().simulate('click');
    expect(onSort.mock.calls[0][0]).toEqual('date')

  });

  it('calls sort method with score option', () => {

    let onSort = jest.fn();

    wrapper = shallow(
      <Headline
        title="All Posts"
        onSort={onSort}
        sortOption="date"
      />
    );
    const component = wrapper.dive();
    component.find(MenuItem).first().simulate('click');
    expect(onSort.mock.calls[0][0]).toEqual('score')

  });

});
