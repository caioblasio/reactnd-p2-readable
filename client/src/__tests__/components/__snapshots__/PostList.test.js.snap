// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Headline /> renders a Headline 1`] = `
<Fragment>
  <div
    className="Headline-root-9"
  >
    <div
      className="Headline-title-10 "
    >
      <WithStyles(Typography)
        className="Headline-header-11"
        variant="h5"
      >
        All Posts
      </WithStyles(Typography)>
    </div>
    <div>
      <WithStyles(Button)
        aria-haspopup="true"
        onClick={[Function]}
        size="small"
        variant="contained"
      >
        <pure(SortIcon)
          className="Headline-leftIcon-12"
        />
        Sorted by score
      </WithStyles(Button)>
      <WithStyles(Menu)
        anchorEl={null}
        id="sort-options"
        onClose={[Function]}
        open={false}
      >
        <WithStyles(MenuItem)
          onClick={[Function]}
          selected={true}
        >
          Vote Score
        </WithStyles(MenuItem)>
        <WithStyles(MenuItem)
          onClick={[Function]}
          selected={false}
        >
          Date
        </WithStyles(MenuItem)>
      </WithStyles(Menu)>
    </div>
  </div>
  <WithStyles(Divider) />
</Fragment>
`;

exports[`<Post /> renders a posts 1`] = `
<Card
  className="Post-card-3 post"
  classes={
    Object {
      "root": "MuiCard-root-8",
    }
  }
  raised={false}
>
  <WithStyles(Menu)
    anchorEl={null}
    id="long-menu"
    onClose={[Function]}
    open={false}
  >
    <Link
      className="Post-link-4"
      replace={false}
      to="/edit/8xf0y6ziyjabvozdd253nd"
    >
      <WithStyles(MenuItem)
        onClick={[Function]}
      >
        Edit
      </WithStyles(MenuItem)>
    </Link>
    <WithStyles(MenuItem)
      onClick={[Function]}
    >
      Remove
    </WithStyles(MenuItem)>
  </WithStyles(Menu)>
  <Link
    className="Post-link-4"
    replace={false}
    to="/categoryone/8xf0y6ziyjabvozdd253nd"
  >
    <WithStyles(CardHeader)
      action={
        <WithStyles(IconButton)
          aria-haspopup="true"
          aria-label="More"
          aria-owns={null}
          onClick={[Function]}
        >
          <pure(MoreVertIcon) />
        </WithStyles(IconButton)>
      }
      subheader="by thingtwo Jun 28, 2016 categoryone"
      title="Udacity is the best place to learn React"
    />
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        component="p"
      >
        Everyone says so after all....
      </WithStyles(Typography)>
    </WithStyles(CardContent)>
  </Link>
  <div
    className="Post-footer-7"
  >
    <WithStyles(Divider) />
    <WithStyles(CardActions)
      className="Post-actions-5"
      disableActionSpacing={true}
    >
      <div
        className="Post-actionIcon-6"
        style={
          Object {
            "flexGrow": 1,
          }
        }
      >
        <Connect(VoteControl)
          id="8xf0y6ziyjabvozdd253nd"
          type="post"
          voteScore={6}
        />
      </div>
      <div
        className="Post-actionIcon-6"
      >
        <FavoriteControl
          id="8xf0y6ziyjabvozdd253nd"
        />
        <WithStyles(IconButton)
          aria-label="Comments"
        >
          <pure(CommentIcon) />
        </WithStyles(IconButton)>
        <WithStyles(Typography)
          component="p"
        >
          2
        </WithStyles(Typography)>
      </div>
    </WithStyles(CardActions)>
  </div>
</Card>
`;
